### ISSUE TITLE
MME output glitches when suggestedLatency <35ms

### ISSUE THREAD

**Issue by dechamps** [PortAudio/portaudio#788]
On my system, `paloopback` reports glitches when outputting through the MME Host API with `suggestedLatency` of 30 ms or less.

This can be reproduced using `paloopback --outputLatency 25` and specifying an MME output device.

Analysis of a 5-second test signal reveals multiple glitches:

![image](https://user-images.githubusercontent.com/866859/224491213-5c9f75c6-2ea1-4901-850c-a6386ae93189.png)

The glitches take the form of small periods of silence (10-50 samples at 48 kHz) being inserted in the stream:

![image](https://user-images.githubusercontent.com/866859/224491423-b5f456f2-58da-48ea-9260-3935c7f06b4e.png)

Now it's quite possible this is "working as intended", as it is completely normal and expected that PortAudio would start to glitch when specifying a suggested latency that is below what the Host API is capable of on a given system. However, in this particular case I have a suspicion that this is not normal. With a 25 ms suggested latency, at 48 kHz, PortAudio allocates 8 MME buffers of 176 samples each. This means the total host buffer size is ~29.3 ms. Investigation reveals that the MME event fires at most every 10 ms, and when it does, between 0 and 3 MME buffers are in the `WHDR_DONE` state - that seems perfectly fine and we never get anywhere close to the point where MME is starved of buffers.

It's possible that this is some kind of MME limitation - if so, maybe that can be worked around by reducing the number of MME buffers.

Otherwise, this could be some kind of PortAudio bug that is only triggered with small MME buffer sizes. For example, it could be that there is a subtle issue when more than one buffer enters the `WHDR_DONE` state in a single `ProcessingThreadProc()` loop iteration - a scenario that is unlikely to happen with large buffers, but is commonplace with small buffers.

- PortAudio version: latest master (b72559b9bc6173ac7209f4ca476b54e8afa73ef4)
- Windows 11 22H2 22621.1265
- Loopback audio device: [Virtual Audio Cable](https://vac.muzychenko.net/en/) 4.67.0.12554

**Comment by dechamps** [comment]
I couldn't find anything obviously wrong on the PortAudio side - it seems to be filling out buffers as soon as they enter `WHDR_DONE` state and in the proper order. Switching from event-driven to ~1ms polling doesn't change anything so it's not an event-related issue.

This really smells like something going wrong on the MME side. Time to pull out our old friend [API Monitor](http://www.rohitab.com/apimonitor) and do some reverse engineering of MME internals:

![image](https://user-images.githubusercontent.com/866859/224543833-59883569-32a8-4b17-b52e-4ba23a686f3f.png)

The above left me scratching me head somewhat. The start (priming) part looks exactly as one would expect: PortAudio supplies 8 buffers and MME dutifully puts all of them in the WASAPI buffer before starting. WASAPI then wakes the MME thread every 10 ms (which is the typical periodicity of the Windows Audio Engine).

Now, what I would expect the MME thread to do is to mark MME buffers as done (and signal the PortAudio thread) in lockstep with the WASAPI buffer being consumed, thus keeping the WASAPI buffer topped off, i.e. targetting 8 MME buffers worth of "padding". But that's not what it's doing. Instead, it lets the WASAPI buffer empty itself off completely, doing nothing for about 20 ms. Then it decides to mark a couple of MME buffers as done and send them back to WASAPI, and then keeps doing that, sending a slow trickle of MME buffers and keeping about 1-3 MME buffers of padding in the WASAPI buffer. Problem is, that is not sufficient padding for smooth playback - WASAPI periodicity is 10 ms, and in that setup a single MME buffer holds ~3.7 ms of data, so clearly this not going to work. And indeed, examination of `GetCurrentPadding()` calls show that the WASAPI buffer underruns on a regular basis in this setup.

So in summary: the problem is that, at any given time, MME is holding on to most of the MME buffers instead of feeding them to WASAPI. MME withholding buffers is causing starvation in the WASAPI buffer, hence the glitches.

PortAudio appears to be innocent here - all MME buffers are filled at all times, it's MME that's not copying the buffers to its WASAPI backend buffer for some reason that I don't quite understand (yet). I'm not sure there is anything that PortAudio can do about this, but perhaps there is way to trick MME into behaving properly.

**Comment by dechamps** [comment]
If I experiment with setting `PA_MME_TARGET_HOST_BUFFER_COUNT_` to 2, again with output suggested latency of 25 ms, PortAudio allocates 2 MME buffers of 25 ms each (which makes sense). But despite the seemingly very wide margins in this case (25 ms is more than twice the event period, and the total buffer size is a very comfortable 50 ms), the exact same scenario plays out:

![image](https://user-images.githubusercontent.com/866859/224560615-06614094-77ad-4c57-b78c-eeab133acf01.png)

MME initially fills the buffer with 50 ms of data before starting the stream, and then for no apparent reason it just sits around doing nothing for the next 40 ms (despite being woken up every 10 ms), resulting in the WASAPI buffer draining completely (underrun). Afterwards it feeds just one buffer at a time instead of double buffering, resulting in repeated underruns.

**Comment by philburk** [comment]
@dechamps - Thanks for the detailed analysis.

> Loopback audio device: [Virtual Audio Cable](https://vac.muzychenko.net/en/) 4.67.0.12554

Have you tried this with a physical USB audio interface with loopback patch cables?

**Comment by dechamps** [comment]
> Have you tried this with a physical USB audio interface with loopback patch cables?

I gave it a try and observed similar behaviour with a variety of hardware, although the threshold seems to change a bit depending on the device. It's possible different devices might respond differently to WASAPI buffer underruns.

I took the liberty of writing a [minimal standalone reproducer](https://github.com/dechamps/MMEBufferSizes) for this issue, which provides additional evidence that the bug is on the MME side. As far as I can tell PortAudio is innocent here, so I'm going to close this. This report can serve as documentation for anyone wondering why seemingly reasonable MME buffer sizes don't work.
